analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs

```diff
@@ -236,21 +236,23 @@ private IEnumerable<ExplodedNode> ExceptionSuccessors(ExplodedNode node, Excepti
         state = node.Block.EnclosingRegion(ControlFlowRegionKind.Catch) is not null
             ? state.PushException(exception)        // If we're nested inside catch, we need to preserve the exception chain
             : state.SetException(exception);        // Otherwise we track only the current exception to avoid explosion of states after each statement
-        while (tryRegion is not null)
+
+        var handlersAreExhaustive = false;
+        while (tryRegion is not null && !handlersAreExhaustive)
         {
             var reachableHandlers = tryRegion.EnclosingRegion.NestedRegions.Where(x => x.Kind != ControlFlowRegionKind.Try && IsReachable(x, state.Exception)).ToArray();
             foreach (var handler in reachableHandlers)  // CatchRegion, FinallyRegion or FilterAndHandlerRegion
             {
+                handlersAreExhaustive |= CatchesAll(handler) || state.Exception.Type.DerivesFrom(handler.ExceptionType);
```

Comment:
For the `DerivesFrom`: What if it's a `FilterAndHandlerRegion` and the condition is false?

Reply:
Good catch. It got lost during refactoring. I'll add a test to make sure it doesn't get lost again.